src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -516,6 +530,43 @@ std::unordered_set<std::size_t> getBloomFilterFilteringColumnKeys(const KeyCondi
     return column_keys;
 }
 
+const parquet::ColumnDescriptor * getColumnDescriptorIfBloomFilterIsPresent(
+    const std::unique_ptr<parquet::RowGroupMetaData> & parquet_rg_metadata,
+    const std::vector<ArrowFieldIndexUtil::ClickHouseIndexToParquetIndex> & clickhouse_column_index_to_parquet_index,
+    std::size_t clickhouse_column_index)
+{
+    if (clickhouse_column_index_to_parquet_index.size() <= clickhouse_column_index)
+    {
+        return nullptr;
+    }
+
+    const auto & parquet_indexes = clickhouse_column_index_to_parquet_index[clickhouse_column_index].parquet_indexes;
+
+    // complex types like structs, tuples and maps will have more than one index.
+    // we don't support those for now
+    if (parquet_indexes.size() > 1)
+    {
+        return nullptr;
+    }
+
+    if (parquet_indexes.empty())
```

Comment:
It crashes regardless of `input_format_parquet_bloom_filter_push_down`.

```
arthur :) select * from file('a.parquet') where x = () settings input_format_parquet_bloom_filter_push_down=0;

SELECT *
FROM file('a.parquet')
WHERE x = tuple()
SETTINGS input_format_parquet_bloom_filter_push_down = 0

Query id: d055ec76-fce2-45bf-8086-0905a3d95348

[arthur] 2024.12.05 10:34:32.673574 [ 2376871 ] {d055ec76-fce2-45bf-8086-0905a3d95348} <Fatal> : Logical error: 'This function cannot be used to construct empty tuple. It is a bug'.
[arthur] 2024.12.05 10:34:32.926912 [ 2376871 ] {d055ec76-fce2-45bf-8086-0905a3d95348} <Fatal> : Stack trace (when copying this message, always include the lines below):

0. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/exception:141: std::exception::capture() @ 0x0000000012f7ef62
1. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/exception:114: std::exception::exception[abi:v15007]() @ 0x0000000012f7ef1b
2. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x0000000049e2ce40
3. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000024905294
4. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:110: DB::Exception::Exception(String&&, int, bool) @ 0x0000000012f67bf7
5. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:63: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000012f61238
6. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:128: DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000012f72d06
7. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Columns/ColumnTuple.cpp:69: DB::ColumnTuple::create(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>> const&) @ 0x000000003aa41636
8. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Columns/ColumnTuple.h:43: DB::ColumnTuple::create(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&&) @ 0x000000001304fb80
9. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:998: DB::readNonNullableColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x000000003e38b735
10. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1142: DB::readColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x000000003e38476a
11. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1198: DB::ArrowColumnToCHColumn::arrowSchemaToCHHeader(arrow::Schema const&, String const&, bool, bool) @ 0x000000003e3830f3
12. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:1173: DB::ParquetSchemaReader::readSchema() @ 0x000000003e8d6c58
13. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Formats/ReadSchemaUtils.cpp:249: DB::readSchemaFromFormatImpl(std::optional<String>, std::optional<DB::FormatSettings> const&, DB::IReadBufferIterator&, std::shared_ptr<DB::Context const> const&) @ 0x000000003e1d8c9a
14. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Formats/ReadSchemaUtils.cpp:549: DB::readSchemaFromFormat(String const&, std::optional<DB::FormatSettings> const&, DB::IReadBufferIterator&, std::shared_ptr<DB::Context const> const&) @ 0x000000003e1defc4
15. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:979: DB::StorageFile::getTableStructureAndFormatFromFileImpl(std::optional<String>, std::vector<String, std::allocator<String>> const&, String const&, std::optional<DB::FormatSettings> const&, std::shared_ptr<DB::Context const> const&, std::optional<DB::StorageFile::ArchiveInfo> const&) @ 0x000000003b405d36
16. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:992: DB::StorageFile::getTableStructureFromFile(String const&, std::vector<String, std::allocator<String>> const&, String const&, std::optional<DB::FormatSettings> const&, std::shared_ptr<DB::Context const> const&, std::optional<DB::StorageFile::ArchiveInfo> const&) @ 0x000000003b406b41
17. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1116: DB::StorageFile::setStorageMetadata(DB::StorageFile::CommonArguments) @ 0x000000003b4095fb
18. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1057: DB::StorageFile::StorageFile(String const&, String const&, DB::StorageFile::CommonArguments) @ 0x000000003b40ab58
19. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1061: DB::StorageFile::StorageFile(String const&, String const&, bool, DB::StorageFile::CommonArguments) @ 0x000000003b40b5df
20. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: DB::StorageFile* std::construct_at[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, DB::StorageFile*>(DB::StorageFile*, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319eac26
21. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298: void std::allocator_traits<std::allocator<DB::StorageFile>>::construct[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, void, void>(std::allocator<DB::StorageFile>&, DB::StorageFile*, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319eaa39
22. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:292: std::__shared_ptr_emplace<DB::StorageFile, std::allocator<DB::StorageFile>>::__shared_ptr_emplace[abi:v15007]<String const&, String, bool, DB::StorageFile::CommonArguments&>(std::allocator<DB::StorageFile>, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319ea8fc
23. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953: std::shared_ptr<DB::StorageFile> std::allocate_shared[abi:v15007]<DB::StorageFile, std::allocator<DB::StorageFile>, String const&, String, bool, DB::StorageFile::CommonArguments&, void>(std::allocator<DB::StorageFile> const&, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319ea60d
24. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962: std::shared_ptr<DB::StorageFile> std::make_shared[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, void>(String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319e7065
25. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/TableFunctionFile.cpp:97: DB::TableFunctionFile::getStorage(String const&, String const&, DB::ColumnsDescription const&, std::shared_ptr<DB::Context const>, String const&, String const&) const @ 0x00000000319e58e7
26. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/ITableFunctionFileLike.cpp:133: DB::ITableFunctionFileLike::executeImpl(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool) const @ 0x00000000319efbe1
27. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/ITableFunction.cpp:36: DB::ITableFunction::execute(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool, bool) const @ 0x0000000032fb081e
28. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Interpreters/Context.cpp:2335: DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::ITableFunction> const&) @ 0x0000000036c5838c
29. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:4870: DB::QueryAnalyzer::resolveTableFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&, bool) @ 0x00000000366ed3a8
30. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:5211: DB::QueryAnalyzer::resolveQueryJoinTreeNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&) @ 0x0000000036728113
31. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:5459: DB::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::IdentifierResolveScope&) @ 0x00000000366da360

[arthur] 2024.12.05 10:35:04.203213 [ 2378198 ] <Fatal> BaseDaemon: ########################################
[arthur] 2024.12.05 10:35:04.203906 [ 2378198 ] <Fatal> BaseDaemon: (version 24.11.1.1, build id: 462772DA68A0DAFC39E1106BF49A9C54965CAB4E, git hash: b76fa9d6af766e49a67fdeea25562148f28402f9) (from thread 2376871) (query_id: d055ec76-fce2-45bf-8086-0905a3d95348) (query: select * from file('a.parquet') where x = () settings input_format_parquet_bloom_filter_push_down=0;) Received signal Aborted (6)
[arthur] 2024.12.05 10:35:04.204267 [ 2378198 ] <Fatal> BaseDaemon: 
[arthur] 2024.12.05 10:35:04.204647 [ 2378198 ] <Fatal> BaseDaemon: Stack trace: 0x0000555579f3a12a 0x0000555579f39f47 0x000055557a9614d9 0x00007ffff7c45320 0x00007ffff7c9eb1d 0x00007ffff7c4526e 0x00007ffff7c288ff 0x0000555579e581f4 0x0000555579e584f0 0x0000555579e59476 0x00005555684bbbf7 0x00005555684b5238 0x00005555684c6d06 0x000055558ff95636 0x00005555685a3b80 0x00005555938df735 0x00005555938d876a 0x00005555938d70f3 0x0000555593e2ac58 0x000055559372cc9a 0x0000555593732fc4 0x0000555590959d36 0x000055559095ab41 0x000055559095d5fb 0x000055559095eb58 0x000055559095f5df 0x0000555586f3ec26 0x0000555586f3ea39 0x0000555586f3e8fc 0x0000555586f3e60d 0x0000555586f3b065 0x0000555586f398e7 0x0000555586f43be1 0x000055558850481e 0x000055558c1ac38c 0x000055558bc413a8 0x000055558bc7c113 0x000055558bc2e360 0x000055558bc2b891 0x000055558bc2ad92 0x000055558d351611 0x000055558dbe202c 0x000055558dbdf9b2 0x000055558dbe6515 0x000055558dbe57c3
[arthur] 2024.12.05 10:35:04.306967 [ 2378198 ] <Fatal> BaseDaemon: 0. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/StackTrace.cpp:380: StackTrace::tryCapture() @ 0x00000000249e612a
[arthur] 2024.12.05 10:35:04.329573 [ 2378198 ] <Fatal> BaseDaemon: 1. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/StackTrace.cpp:349: StackTrace::StackTrace(ucontext_t const&) @ 0x00000000249e5f47
[arthur] 2024.12.05 10:35:04.392383 [ 2378198 ] <Fatal> BaseDaemon: 2. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/SignalHandlers.cpp:85: signalHandler(int, siginfo_t*, void*) @ 0x000000002540d4d9
[arthur] 2024.12.05 10:35:04.392567 [ 2378198 ] <Fatal> BaseDaemon: 3. ? @ 0x00007ffff7c45320
[arthur] 2024.12.05 10:35:04.392689 [ 2378198 ] <Fatal> BaseDaemon: 4. ? @ 0x00007ffff7c9eb1d
[arthur] 2024.12.05 10:35:04.392785 [ 2378198 ] <Fatal> BaseDaemon: 5. ? @ 0x00007ffff7c4526e
[arthur] 2024.12.05 10:35:04.392911 [ 2378198 ] <Fatal> BaseDaemon: 6. ? @ 0x00007ffff7c288ff
[arthur] 2024.12.05 10:35:04.457047 [ 2378198 ] <Fatal> BaseDaemon: 7. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.cpp:48: DB::abortOnFailedAssertion(String const&, void* const*, unsigned long, unsigned long) @ 0x00000000249041f4
[arthur] 2024.12.05 10:35:04.480793 [ 2378198 ] <Fatal> BaseDaemon: 8. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.cpp:71: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&) @ 0x00000000249044f0
[arthur] 2024.12.05 10:35:04.502878 [ 2378198 ] <Fatal> BaseDaemon: 9. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.cpp:115: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000024905476
[arthur] 2024.12.05 10:35:04.518807 [ 2378198 ] <Fatal> BaseDaemon: 10. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:110: DB::Exception::Exception(String&&, int, bool) @ 0x0000000012f67bf7
[arthur] 2024.12.05 10:35:04.536120 [ 2378198 ] <Fatal> BaseDaemon: 11. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:63: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000012f61238
[arthur] 2024.12.05 10:35:04.552525 [ 2378198 ] <Fatal> BaseDaemon: 12. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Common/Exception.h:128: DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000012f72d06
[arthur] 2024.12.05 10:35:04.667522 [ 2378198 ] <Fatal> BaseDaemon: 13. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Columns/ColumnTuple.cpp:69: DB::ColumnTuple::create(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>> const&) @ 0x000000003aa41636
[arthur] 2024.12.05 10:35:04.681705 [ 2378198 ] <Fatal> BaseDaemon: 14. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Columns/ColumnTuple.h:43: DB::ColumnTuple::create(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&&) @ 0x000000001304fb80
[arthur] 2024.12.05 10:35:04.879637 [ 2378198 ] <Fatal> BaseDaemon: 15. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:998: DB::readNonNullableColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x000000003e38b735
[arthur] 2024.12.05 10:35:04.981558 [ 2378198 ] <Fatal> BaseDaemon: 16. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1142: DB::readColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x000000003e38476a
[arthur] 2024.12.05 10:35:05.116591 [ 2378198 ] <Fatal> BaseDaemon: 17. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1198: DB::ArrowColumnToCHColumn::arrowSchemaToCHHeader(arrow::Schema const&, String const&, bool, bool) @ 0x000000003e3830f3
[arthur] 2024.12.05 10:35:05.565617 [ 2378198 ] <Fatal> BaseDaemon: 18. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:1173: DB::ParquetSchemaReader::readSchema() @ 0x000000003e8d6c58
[arthur] 2024.12.05 10:35:05.709858 [ 2378198 ] <Fatal> BaseDaemon: 19. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Formats/ReadSchemaUtils.cpp:249: DB::readSchemaFromFormatImpl(std::optional<String>, std::optional<DB::FormatSettings> const&, DB::IReadBufferIterator&, std::shared_ptr<DB::Context const> const&) @ 0x000000003e1d8c9a
[arthur] 2024.12.05 10:35:05.812525 [ 2378198 ] <Fatal> BaseDaemon: 20. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Formats/ReadSchemaUtils.cpp:549: DB::readSchemaFromFormat(String const&, std::optional<DB::FormatSettings> const&, DB::IReadBufferIterator&, std::shared_ptr<DB::Context const> const&) @ 0x000000003e1defc4
[arthur] 2024.12.05 10:35:06.128309 [ 2378198 ] <Fatal> BaseDaemon: 21. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:979: DB::StorageFile::getTableStructureAndFormatFromFileImpl(std::optional<String>, std::vector<String, std::allocator<String>> const&, String const&, std::optional<DB::FormatSettings> const&, std::shared_ptr<DB::Context const> const&, std::optional<DB::StorageFile::ArchiveInfo> const&) @ 0x000000003b405d36
[arthur] 2024.12.05 10:35:06.263744 [ 2378198 ] <Fatal> BaseDaemon: 22. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:992: DB::StorageFile::getTableStructureFromFile(String const&, std::vector<String, std::allocator<String>> const&, String const&, std::optional<DB::FormatSettings> const&, std::shared_ptr<DB::Context const> const&, std::optional<DB::StorageFile::ArchiveInfo> const&) @ 0x000000003b406b41
[arthur] 2024.12.05 10:35:06.363960 [ 2378198 ] <Fatal> BaseDaemon: 23. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1116: DB::StorageFile::setStorageMetadata(DB::StorageFile::CommonArguments) @ 0x000000003b4095fb
[arthur] 2024.12.05 10:35:06.485913 [ 2378198 ] <Fatal> BaseDaemon: 24. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1057: DB::StorageFile::StorageFile(String const&, String const&, DB::StorageFile::CommonArguments) @ 0x000000003b40ab58
[arthur] 2024.12.05 10:35:06.598660 [ 2378198 ] <Fatal> BaseDaemon: 25. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Storages/StorageFile.cpp:1061: DB::StorageFile::StorageFile(String const&, String const&, bool, DB::StorageFile::CommonArguments) @ 0x000000003b40b5df
[arthur] 2024.12.05 10:35:06.650744 [ 2378198 ] <Fatal> BaseDaemon: 26. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: DB::StorageFile* std::construct_at[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, DB::StorageFile*>(DB::StorageFile*, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319eac26
[arthur] 2024.12.05 10:35:06.696908 [ 2378198 ] <Fatal> BaseDaemon: 27. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298: void std::allocator_traits<std::allocator<DB::StorageFile>>::construct[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, void, void>(std::allocator<DB::StorageFile>&, DB::StorageFile*, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319eaa39
[arthur] 2024.12.05 10:35:06.743266 [ 2378198 ] <Fatal> BaseDaemon: 28. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:292: std::__shared_ptr_emplace<DB::StorageFile, std::allocator<DB::StorageFile>>::__shared_ptr_emplace[abi:v15007]<String const&, String, bool, DB::StorageFile::CommonArguments&>(std::allocator<DB::StorageFile>, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319ea8fc
[arthur] 2024.12.05 10:35:06.789106 [ 2378198 ] <Fatal> BaseDaemon: 29. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953: std::shared_ptr<DB::StorageFile> std::allocate_shared[abi:v15007]<DB::StorageFile, std::allocator<DB::StorageFile>, String const&, String, bool, DB::StorageFile::CommonArguments&, void>(std::allocator<DB::StorageFile> const&, String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319ea60d
[arthur] 2024.12.05 10:35:06.836075 [ 2378198 ] <Fatal> BaseDaemon: 30. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962: std::shared_ptr<DB::StorageFile> std::make_shared[abi:v15007]<DB::StorageFile, String const&, String, bool, DB::StorageFile::CommonArguments&, void>(String const&, String&&, bool&&, DB::StorageFile::CommonArguments&) @ 0x00000000319e7065
[arthur] 2024.12.05 10:35:06.880606 [ 2378198 ] <Fatal> BaseDaemon: 31. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/TableFunctionFile.cpp:97: DB::TableFunctionFile::getStorage(String const&, String const&, DB::ColumnsDescription const&, std::shared_ptr<DB::Context const>, String const&, String const&) const @ 0x00000000319e58e7
[arthur] 2024.12.05 10:35:06.923148 [ 2378198 ] <Fatal> BaseDaemon: 32. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/ITableFunctionFileLike.cpp:133: DB::ITableFunctionFileLike::executeImpl(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool) const @ 0x00000000319efbe1
[arthur] 2024.12.05 10:35:06.969905 [ 2378198 ] <Fatal> BaseDaemon: 33. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/TableFunctions/ITableFunction.cpp:36: DB::ITableFunction::execute(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool, bool) const @ 0x0000000032fb081e
[arthur] 2024.12.05 10:35:07.230089 [ 2378198 ] <Fatal> BaseDaemon: 34. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Interpreters/Context.cpp:2335: DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::ITableFunction> const&) @ 0x0000000036c5838c
[arthur] 2024.12.05 10:35:07.426200 [ 2378198 ] <Fatal> BaseDaemon: 35. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:4870: DB::QueryAnalyzer::resolveTableFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&, bool) @ 0x00000000366ed3a8
[arthur] 2024.12.05 10:35:07.562739 [ 2378198 ] <Fatal> BaseDaemon: 36. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:5211: DB::QueryAnalyzer::resolveQueryJoinTreeNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&) @ 0x0000000036728113
[arthur] 2024.12.05 10:35:07.716507 [ 2378198 ] <Fatal> BaseDaemon: 37. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:5459: DB::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::IdentifierResolveScope&) @ 0x00000000366da360
[arthur] 2024.12.05 10:35:07.828813 [ 2378198 ] <Fatal> BaseDaemon: 38. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalyzer.cpp:169: DB::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>) @ 0x00000000366d7891
[arthur] 2024.12.05 10:35:07.844850 [ 2378198 ] <Fatal> BaseDaemon: 39. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/Resolve/QueryAnalysisPass.cpp:18: DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x00000000366d6d92
[arthur] 2024.12.05 10:35:07.886570 [ 2378198 ] <Fatal> BaseDaemon: 40. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Analyzer/QueryTreePassManager.cpp:185: DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x0000000037dfd611
[arthur] 2024.12.05 10:35:07.939154 [ 2378198 ] <Fatal> BaseDaemon: 41. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:149: DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) @ 0x000000003868e02c
[arthur] 2024.12.05 10:35:07.972685 [ 2378198 ] <Fatal> BaseDaemon: 42. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:167: DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x000000003868b9b2
[arthur] 2024.12.05 10:35:08.009465 [ 2378198 ] <Fatal> BaseDaemon: 43. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:714: std::__unique_if<DB::InterpreterSelectQueryAnalyzer>::__unique_single std::make_unique[abi:v15007]<DB::InterpreterSelectQueryAnalyzer, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&>(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&) @ 0x0000000038692515
[arthur] 2024.12.05 10:35:08.078389 [ 2378198 ] <Fatal> BaseDaemon: 44. /home/laptop/work/merge_parquet_minmax_bloom_filter_evaluation/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:279: DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0::operator()(DB::InterpreterFactory::Arguments const&) const @ 0x00000000386917c3
[arthur] 2024.12.05 10:35:08.078542 [ 2378198 ] <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
[arthur] 2024.12.05 10:35:11.492469 [ 2378198 ] <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
[arthur] 2024.12.05 10:35:11.492815 [ 2378198 ] <Fatal> BaseDaemon: Changed settings: log_queries = true, allow_introspection_functions = true, input_format_parquet_bloom_filter_push_down = false

```

Reply:
https://fiddle.clickhouse.com/ee995c63-b66c-4889-be13-38c1a819285c

Reply:
Oh, I almost ended up submitting a PR with the same code as you did in https://github.com/ClickHouse/ClickHouse/pull/72616.

Reply:
With those changes applied, it does not crash.

```
arthur :) select * from file('a.parquet') where x = () settings input_format_parquet_bloom_filter_push_down=1;

SELECT *
FROM file('a.parquet')
WHERE x = tuple()
SETTINGS input_format_parquet_bloom_filter_push_down = 1

Query id: d8b27ed1-dbf6-4cf3-8d0c-fef23f9c5692


Elapsed: 0.010 sec. 

Received exception from server (version 24.11.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: Comparison of zero-sized tuples is not implemented.. (NOT_IMPLEMENTED)

```

Reply:
Right, sorry I forgot that that PR is needed to even test this.

Ok, maybe empty-tuple-typed columns can't show up in KeyCondition right now, so we can't test this. So this `if` will be hit in 3 cases:
 (a) there's a bug,
 (b) it's the future, and empty tuples can appear in KeyCondition (e.g. empty tuple comparison was implemented),
 (c) something else I missed.

Out of these, (a) seems the most likely, so I guess we should keep the LOGICAL_ERROR, no changes. Maybe change "Something bad happened" to something more specific like "Column maps to 0 parquet leaf columns".